server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java

```diff
@@ -305,6 +305,10 @@ protected void search(List<LeafReaderContext> leaves, Weight weight, Collector c
      */
     @Override
     protected void searchLeaf(LeafReaderContext ctx, Weight weight, Collector collector) throws IOException {
+        // If search is already flagged as timed out, do not continue
+        if (searchContext.isSearchTimedOut()) {
```

Comment:
@jed326 @sohami @kkewwei it seems like when we were adapting to concurrent search, the exception driven flow was changed to state driven one, and the leaves are being processed even when the search request has exceeded the time budget (timed out).

Reply:
Thanks @reta, I will have more time to take a closer look at this later this week.

From what I recall off the top of my head the `cancellable.checkCancelled();` should be doing the timeout checking and then throwing the `QueryPhase.TimeExceededException`. The `searchContext.setSearchTimedOut(true);` logic was needed because with concurrent search any exceptions thrown in  `searchLeaf` would be handled by the Lucene `TaskExecutor` and it was difficult to figure out from the query phase thread if the search had actually timed out or not.

Reply:
> Thanks @reta, I will have more time to take a closer look at this later this week.

Thanks @jed326 , I will look closely as well, but `cancellable.checkCancelled()` seems to be somehow not short-circuiting the search, I could clearly see that by checking the how long tests are taking w/o this change.

Reply:
> `cancellable.checkCancelled()` seems to be somehow not short-circuiting the search

That's definitely strange. I do see that `searchLeaf` gets called for each leaf even post the timeout and we leave it to `cancellable.checkCancelled()` to throw an exception which we than catch for each leaf. I wonder if this repeated exception handling is what is causing it to take longer?

Maybe we should even have the `searchContext.isSearchTimedOut()` check in the `search` method above so we can short circuit those for loops when the timeout is hit?

Reply:
> Maybe we should even have the `searchContext.isSearchTimedOut()` check in the `search` method above so we can short circuit those for loops when the timeout is hit?

What I've seen, the timeout is hit by first `searchLeaf` call (those are called in loop), so the check in `search` method may not need this check (I think)

Reply:
> I could clearly see that by checking how long tests are taking w/o this change.

Currently the only reason I can think why this would be the case is because `cancellable.checkCancelled()` involves throwing and catching an exception which is slowing things down. I think we may need to run this piece of code through a profiler to understand why.

Reply:
@jed326 @kkewwei thanks for the feedback folks, will spend more time looking into the cause